modules/distributor/queue/queue.go

```diff
@@ -0,0 +1,225 @@
+	logger log.Logger
+
+	name           string
+	tenantID       string
+	workerCount    int
+	size           int
+	reqChan        chan *request[T]
+	fn             ProcessFunc[T]
+	workersCloseCh chan struct{}
+
+	pushesTotalMetrics         *prometheus.CounterVec
+	pushesFailuresTotalMetrics *prometheus.CounterVec
+	lengthMetric               *prometheus.GaugeVec
+
+	readOnly *atomic.Bool
+}
+
+func New[T any](cfg Config, logger log.Logger, reg prometheus.Registerer, fn ProcessFunc[T]) *Queue[T] {
+	pushesTotalMetrics := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_total",
+		Help:      "Total number of successful requests queued up for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	pushesFailuresTotalMetric := prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_pushes_failures_total",
+		Help:      "Total number of failed pushes to the queue for a tenant to the generatorForwarder",
+	}, []string{"name", "tenant"})
+	lengthMetric := prometheus.NewGaugeVec(prometheus.GaugeOpts{
+		Namespace: "tempo",
+		Subsystem: "distributor",
+		Name:      "queue_length",
+		Help:      "Number of queued requests for a tenant",
+	}, []string{"name", "tenant"})
+
+	var alreadyRegisteredErr prometheus.AlreadyRegisteredError
+	err := reg.Register(pushesTotalMetrics)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_total metric", "err", err)
+	}
+	err = reg.Register(pushesFailuresTotalMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_pushes_failures_total metric", "err", err)
+	}
+
+	err = reg.Register(lengthMetric)
+	if err != nil && !errors.As(err, &alreadyRegisteredErr) {
+		_ = level.Warn(logger).Log("msg", "failed to register queue_length metric", "err", err)
+	}
+
+	return &Queue[T]{
+		logger:                     logger,
+		name:                       cfg.Name,
+		tenantID:                   cfg.TenantID,
+		workerCount:                cfg.WorkerCount,
+		size:                       cfg.Size,
+		reqChan:                    make(chan *request[T], cfg.Size),
+		fn:                         fn,
+		workersCloseCh:             make(chan struct{}),
+		pushesTotalMetrics:         pushesTotalMetrics,
+		pushesFailuresTotalMetrics: pushesFailuresTotalMetric,
+		lengthMetric:               lengthMetric,
+		readOnly:                   atomic.NewBool(false),
+	}
+}
+
+// Push pushes data onto a queue.
+// If the queue is full, the data is dropped
+func (m *Queue[T]) Push(ctx context.Context, data T) error {
+	if m.readOnly.Load() {
+		return fmt.Errorf("queue is read-only")
+	}
+
+	m.pushesTotalMetrics.WithLabelValues(m.name, m.tenantID).Inc()
+
+	reqCtx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
```

Comment:
The core idea was to remove `context.Background` that was previously used in queue manager, like this:

```go
m.forwardRequest(context.Background(), req)
```

and add at least some timeout. I've put it here because it's now easy to derive this context from the context that we receive in Push() as an argument if need be. I don't have strong opinions on this, though.

Reply:
I know understand what you meant when you wrote "context is canceled". You're right, that's not the intended behaviour. I fixed it and moved the context back where it was before (to the queue itself). However, I kept the 10s timeout :)